### PR TITLE
bump mypy and fix related new errors

### PR DESCRIPTION
**Pull request by pacrob**
### What was wrong?

`mypy` and `mypy_extensions` versions were holding back typing updates elsewhere.

### How was it fixed?

Bumped! Fixed new errors where I could, was unable to resolve typing in 2 spots.

### Todo:

- [x] Add entry to the [release notes](https://github.com/ethereum/web3.py/blob/master/newsfragments/README.md)

#### Cute Animal Picture

image

### CODE REVIEW

**Review by reedsa — APPROVED** [review]
lgtm!

**Review by fselmo — COMMENTED** [review]
Looks good overall. I think there is some possible type-narrowing we can do.

File: eth/abc.py [hunk 1]
```diff
@@ -487,7 +487,9 @@ class UnsignedTransactionAPI(BaseTransactionAPI):
     # API that must be implemented by all Transaction subclasses.
     #
     @abstractmethod
-    def as_signed_transaction(self, private_key: PrivateKey) -> "SignedTransactionAPI":
+    def as_signed_transaction(
+        self, private_key: PrivateKey, chain_id: int
```
[the remark this answers is not in the document]
Comment (pacrob) [reply eth/abc.py:491]: I wasn't sure about defining a default in an abstract method, but yes!

**Review by fselmo — APPROVED** [review]
lgtm after tests pass 👌🏼

**Pull request merged**